### PR TITLE
avoid some useless warning or error

### PR DESCRIPTION
**Pull request by xsren**
When I run **general-spider** in the examples, I get some errors and warnings. After reading the source code, I find that they did not affect the result,  but they are really distrubing and misleading, especially to newcomer like me. So I make this pull request.

The warning:
```
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
2017-02-16 11:48:05 [frontera.contrib.backends.remote.codecs.msgpack] WARNING: unable to serialize object: None
......
```

The error:
```
 2017-02-16 10:33:54,650 /Library/Python/2.7/site-packages/frontera/worker/db.py error[line:48] ERROR    db-worker       0
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/twisted/internet/defer.py", line 587, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/Library/Python/2.7/site-packages/frontera/utils/async.py", line 41, in __call__
    return self._func(*self._a, **self._kw)
  File "/Library/Python/2.7/site-packages/frontera/worker/db.py", line 172, in consume_incoming
    producer_offset = self.spider_feed_producer.get_offset(partition_id)
  File "/Library/Python/2.7/site-packages/frontera/contrib/messagebus/zeromq/__init__.py", line 103, in get_offset
    return self.counters[partition_id]
KeyError: 0
Unhandled error in Deferred:


Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/twisted/internet/base.py", line 1204, in mainLoop
    self.runUntilCurrent()
  File "/Library/Python/2.7/site-packages/twisted/internet/base.py", line 825, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/Library/Python/2.7/site-packages/twisted/internet/defer.py", line 393, in callback
    self._startRunCallbacks(result)
  File "/Library/Python/2.7/site-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/Library/Python/2.7/site-packages/twisted/internet/defer.py", line 587, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/Library/Python/2.7/site-packages/frontera/utils/async.py", line 41, in __call__
    return self._func(*self._a, **self._kw)
  File "/Library/Python/2.7/site-packages/frontera/worker/db.py", line 172, in consume_incoming
    producer_offset = self.spider_feed_producer.get_offset(partition_id)
  File "/Library/Python/2.7/site-packages/frontera/contrib/messagebus/zeromq/__init__.py", line 103, in get_offset
    return self.counters[partition_id]
exceptions.KeyError: 0
```

### CODE REVIEW

File: frontera/contrib/messagebus/zeromq/__init__.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def flush(self):
         pass
 
     def get_offset(self, partition_id):
-        return self.counters[partition_id]
+        return self.counters.get(partition_id,0)
```
Comment (voith) [inline frontera/contrib/messagebus/zeromq/__init__.py]: Can you please leave a whitespace after the comma in order to pass the failing flake8 tests

File: frontera/contrib/backends/remote/codecs/msgpack.py [hunk 2]
```diff
@@ -30,7 +30,8 @@ def serialize(obj):
         elif hasattr(obj, '__dict__'):
             return serialize(obj.__dict__)
         else:
-            logger.warning('unable to serialize object: {}'.format(obj))
+            if obj:
```
Comment (voith) [inline frontera/contrib/backends/remote/codecs/msgpack.py]: This logging problem is in fact a bug added by me in this [PR](https://github.com/scrapinghub/frontera/pull/233/files#diff-c2fba22f948d4da57c25c9c26263a9b7R33). Thank you for this! 

However your fix doesn't seem right to me. What happens when `obj=True` or say there is a key in a dict which has a value that is True? It will still output the warning message in such a case. 

In fact now If I think about this whole method looks buggy to me. Whenever a dict contains booleans which can be easily serialized this method will convert them into None. 

This method made more sense when it would encode string objects to `utf-8`. But this was removed [here](https://github.com/scrapinghub/frontera/pull/233/files#diff-c2fba22f948d4da57c25c9c26263a9b7R18) as part of the fix. 
What this method really does of value  now is that it [copies](https://github.com/scrapinghub/frontera/blob/master/frontera/contrib/backends/remote/codecs/msgpack.py#L22) the dict. But IMO, this was needed when strings were forced to be encoded to `utf-8`. I no longer feel that this method is needed. 
If you're really interested in fixing this than you can try removing this method completely and test it by adding some unit tests.

File: frontera/contrib/messagebus/zeromq/__init__.py (hunk 1 above)
Comment (xsren) [inline frontera/contrib/messagebus/zeromq/__init__.py]: OK!

File: frontera/contrib/backends/remote/codecs/msgpack.py (hunk 2 above)
Comment (xsren) [inline frontera/contrib/backends/remote/codecs/msgpack.py]: Thanks for your reply!  But I still have some question?

> However your fix doesn't seem right to me. What happens when obj=True or say there is a key in a dict which has a value that is True? It will still output the warning message in such a case.

When _obj=True_ , isn't the program going to this condition  ? 
```
if isinstance(obj, (bool, six.integer_types, float, six.binary_type, six.text_type)):
    return obj
```

Another question, will you not fix this bug?

Comment (voith) [inline frontera/contrib/backends/remote/codecs/msgpack.py]: > if isinstance(obj, (bool, six.integer_types, float, six.binary_type, six.text_type)):
            return obj

You are right about the boolean part. did not see that. thanks!
> Another question, will you not fix this bug?

Fair enough. I will fix it myself! Hopefully tonight(IST).
I have to verify if the `serialize` method is really needed.

Comment (xsren) [inline frontera/contrib/backends/remote/codecs/msgpack.py]: Ok! Thank you!

**Comment by sibiryakov** [comment]
Oh guys. I recommend to fix the **cause** of error, not it's consequences. Why key is absent in this dict? Where it should be set?

**Comment by voith** [comment]
Sorry guys I didn't really get time to look into this. 
> I recommend to fix the cause of error, not it's consequences.

I agree with you that something wrong must be going on. But this issue of logging must be definitely be fixed. A user can always pass `None` in scrapy meta but this shouldn't be logged as a warning, IMO

**Comment by sibiryakov** [comment]
IMO it would be more readable to introduce another ``elif`` clause checking for NoneType instead of adding conditional logging.

File: frontera/contrib/messagebus/zeromq/__init__.py [hunk 3]
```diff
@@ -98,7 +98,7 @@ def flush(self):
         pass
 
     def get_offset(self, partition_id):
-        return self.counters[partition_id]
+        return self.counters.get(partition_id, 0)
```
Comment (sibiryakov) [inline frontera/contrib/messagebus/zeromq/__init__.py]: why is this needed? Looks like a hot fix.

Comment (xsren) [inline frontera/contrib/messagebus/zeromq/__init__.py]: When I run **general-spider** in the examples, it output error:

```
 2017-02-16 10:33:54,650 /Library/Python/2.7/site-packages/frontera/worker/db.py error[line:48] ERROR    db-worker       0
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/twisted/internet/defer.py", line 587, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/Library/Python/2.7/site-packages/frontera/utils/async.py", line 41, in __call__
    return self._func(*self._a, **self._kw)
  File "/Library/Python/2.7/site-packages/frontera/worker/db.py", line 172, in consume_incoming
    producer_offset = self.spider_feed_producer.get_offset(partition_id)
  File "/Library/Python/2.7/site-packages/frontera/contrib/messagebus/zeromq/__init__.py", line 103, in get_offset
    return self.counters[partition_id]
KeyError: 0
Unhandled error in Deferred:


Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/twisted/internet/base.py", line 1204, in mainLoop
    self.runUntilCurrent()
  File "/Library/Python/2.7/site-packages/twisted/internet/base.py", line 825, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/Library/Python/2.7/site-packages/twisted/internet/defer.py", line 393, in callback
    self._startRunCallbacks(result)
  File "/Library/Python/2.7/site-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/Library/Python/2.7/site-packages/twisted/internet/defer.py", line 587, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/Library/Python/2.7/site-packages/frontera/utils/async.py", line 41, in __call__
    return self._func(*self._a, **self._kw)
  File "/Library/Python/2.7/site-packages/frontera/worker/db.py", line 172, in consume_incoming
    producer_offset = self.spider_feed_producer.get_offset(partition_id)
  File "/Library/Python/2.7/site-packages/frontera/contrib/messagebus/zeromq/__init__.py", line 103, in get_offset
    return self.counters[partition_id]
exceptions.KeyError: 0
```
